--- b/fs/btrfs/file.c	2016-03-23 07:38:02.000000000 -0400
+++ a/fs/btrfs/file.c	2016-03-09 06:49:43.000000000 -0500
@@ -1873,7 +1873,7 @@
 int btrfs_sync_file(struct file *file, loff_t start, loff_t end, int datasync)
 {
 	struct dentry *dentry = file->f_path.dentry;
-	struct inode *inode = dentry->d_inode;
+	struct inode *inode = file_inode(file);
 	struct btrfs_root *root = BTRFS_I(inode)->root;
 	struct btrfs_trans_handle *trans;
 	struct btrfs_log_ctx ctx;
