--- a/modules/cudev/include/opencv2/cudev/util/saturate_cast.hpp	2017-08-03 23:58:23.000000000 +0000
+++ b/modules/cudev/include/opencv2/cudev/util/saturate_cast.hpp	2017-11-28 17:50:56.064790763 +0000
@@ -48,6 +48,10 @@
 
 #include "../common.hpp"
 
+#if __CUDACC_VER_MAJOR__ >= 9
+#include <cuda_fp16.h>
+#endif
+
 namespace cv { namespace cudev {
 
 //! @addtogroup cudev
@@ -274,12 +278,21 @@
 
 template <> __device__ __forceinline__ float cast_fp16<short, float>(short v)
 {
+#if __CUDACC_VER_MAJOR__ >= 9
+    return float(*(__half*)&v);
+#else
     return __half2float(v);
+#endif
 }
 
 template <> __device__ __forceinline__ short cast_fp16<float, short>(float v)
 {
+#if __CUDACC_VER_MAJOR__ >= 9
+    __half h(v);
+    return *(short*)&v;
+#else
     return (short)__float2half_rn(v);
+#endif0
 }
 //! @}
