diff --git a/lustre/utils/lfs.c b/lustre/utils/lfs.c
index be6a2d3f8e..0798a0c876 100644
--- a/lustre/utils/lfs.c
+++ b/lustre/utils/lfs.c
@@ -5194,11 +5194,13 @@ static const char *const ladvise_names[] = LU_LADVISE_NAMES;
 
 static const char *const lock_mode_names[] = LOCK_MODE_NAMES;
 
+#if 0
 static const char *const lockahead_results[] = {
 	[LLA_RESULT_SENT] = "Lock request sent",
 	[LLA_RESULT_DIFFERENT] = "Different matching lock found",
 	[LLA_RESULT_SAME] = "Matching lock on identical extent found",
 };
+#endif
 
 int lfs_get_mode(const char *string)
 {
