--- src/basic/socket-util.c.orig	2016-09-09 08:16:50.120031646 -0400
+++ src/basic/socket-util.c	2016-09-09 08:17:12.751216751 -0400
@@ -818,8 +818,10 @@
                 if ((unsigned char) *p <= 32U)
                         return false;
 
+#if 0
                 if (*p == ':' || *p == '/')
                         return false;
+#endif
 
                 numeric = numeric && (*p >= '0' && *p <= '9');
                 p++;
