--- systemd-220/src/nspawn/nspawn.c.orig	2015-07-16 10:03:04.009377414 +0000
+++ systemd-220/src/nspawn/nspawn.c	2015-07-16 10:11:21.116183708 +0000
@@ -1322,7 +1322,6 @@
 
 static int mount_cgroup(const char *dest) {
         _cleanup_set_free_free_ Set *controllers = NULL;
-        const char *cgroup_root;
         int r;
 
         controllers = set_new(&string_hash_ops);
@@ -1348,7 +1347,7 @@
                 if (r == -EINVAL) {
                         /* Not a symbolic link, but directly a single cgroup hierarchy */
 
-                        r = mount_cgroup_hierarchy(dest, controller, controller, true);
+                        r = mount_cgroup_hierarchy(dest, controller, controller, false);
                         if (r < 0)
                                 return r;
 
@@ -1368,7 +1367,7 @@
                                 continue;
                         }
 
-                        r = mount_cgroup_hierarchy(dest, combined, combined, true);
+                        r = mount_cgroup_hierarchy(dest, combined, combined, false);
                         if (r < 0)
                                 return r;
 
@@ -1386,10 +1385,6 @@
         if (r < 0)
                 return r;
 
-        cgroup_root = prefix_roota(dest, "/sys/fs/cgroup");
-        if (mount(NULL, cgroup_root, NULL, MS_REMOUNT|MS_NOSUID|MS_NOEXEC|MS_NODEV|MS_STRICTATIME|MS_RDONLY, "mode=755") < 0)
-                return log_error_errno(errno, "Failed to remount %s read-only: %m", cgroup_root);
-
         return 0;
 }
 
