diff -u -r -N a/man/resolved.conf.xml b/man/resolved.conf.xml
--- a/man/resolved.conf.xml	2017-03-01 16:43:06.000000000 -0500
+++ b/man/resolved.conf.xml	2017-03-08 07:22:17.756610411 -0500
@@ -225,6 +225,12 @@
         in use.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>mDNS=</varname></term>
+        <listitem><para>Takes a boolean argument. If <literal>yes</literal> (the default), mDNS stack is enabled.
+        If <literal>no</literal>, the mDNS stack is disabled.</para></listitem>
+      </varlistentry>
+
     </variablelist>
   </refsect1>
 
diff -u -r -N a/src/resolve/resolved.conf.in b/src/resolve/resolved.conf.in
--- a/src/resolve/resolved.conf.in	2017-03-01 16:43:06.000000000 -0500
+++ b/src/resolve/resolved.conf.in	2017-03-08 07:21:27.979222145 -0500
@@ -19,3 +19,4 @@
 #DNSSEC=@DEFAULT_DNSSEC_MODE@
 #Cache=yes
 #DNSStubListener=udp
+#mDNS=yes
diff -u -r -N a/src/resolve/resolved-gperf.gperf b/src/resolve/resolved-gperf.gperf
--- a/src/resolve/resolved-gperf.gperf	2017-03-01 16:43:06.000000000 -0500
+++ b/src/resolve/resolved-gperf.gperf	2017-03-08 07:21:06.602055397 -0500
@@ -21,3 +21,4 @@
 Resolve.DNSSEC,          config_parse_dnssec_mode,            0,                   offsetof(Manager, dnssec_mode)
 Resolve.Cache,           config_parse_bool,                   0,                   offsetof(Manager, enable_cache)
 Resolve.DNSStubListener, config_parse_dns_stub_listener_mode, 0,                   offsetof(Manager, dns_stub_listener_mode)
+Resolve.mDNS,            config_parse_bool,                   0,                   offsetof(Manager, mdns_support)
