--- a/libcfs/include/libcfs/linux/linux-time.h	2016-03-23 17:58:25.818000000 +0000
+++ b/libcfs/include/libcfs/linux/linux-time.h	2016-03-23 17:59:00.931000000 +0000
@@ -108,7 +108,9 @@
 typedef cycles_t cfs_cycles_t;
 
 #define timespec64		timespec
+#ifndef ktime_get_real_ts64
 #define ktime_get_real_ts64	ktime_get_real_ts
+#endif
 
 static inline int cfs_time_before(cfs_time_t t1, cfs_time_t t2)
 {
