diff -u -r -N a/src/util/tev_curl.c b/src/util/tev_curl.c
--- a/src/util/tev_curl.c	2018-12-20 15:31:27.493987239 -0500
+++ b/src/util/tev_curl.c	2018-12-20 15:31:39.577796491 -0500
@@ -97,7 +97,9 @@
         return ETIMEDOUT;
     case CURLE_SSL_ISSUER_ERROR:
     case CURLE_SSL_CACERT_BADFILE:
+#if LIBCURL_VERSION_NUM < 0x073e00
     case CURLE_SSL_CACERT:
+#endif
     case CURLE_SSL_CERTPROBLEM:
         return ERR_INVALID_CERT;
 
