--- Collage/co/dataIStreamArchive.h.orig	2016-09-12 08:36:37.000000000 -0400
+++ Collage/co/dataIStreamArchive.h	2016-09-12 08:36:59.000000000 -0400
@@ -72,7 +72,7 @@
 
     /** @internal use optimized load for arrays. */
     template< typename T >
-    void load_array( boost::serialization::array< T >& a, unsigned int );
+    void load_array( boost::serialization::array_wrapper< T >& a, unsigned int );
 
     /** @internal enable serialization optimization for arrays. */
     struct use_array_optimization
