--- a/contrib/init/systemd/docker.service	2016-04-27 12:20:25.490000000 +0000
+++ b/contrib/init/systemd/docker.service	2016-04-27 12:21:25.998000000 +0000
@@ -9,7 +9,7 @@
 # the default is not to use systemd for cgroups because the delegate issues still
 # exists and systemd currently does not support the cgroup feature set required
 # for containers run by docker
-ExecStart=/usr/bin/docker daemon -H fd://
+ExecStart=/usr/bin/dockerd -H fd://
 MountFlags=slave
 LimitNOFILE=1048576
 LimitNPROC=1048576
